### PR TITLE
ocamlPackages.ocamlgraph: 1.8.8 -> 2.0.0

### PR DESCRIPTION
**Pull request by superherointj**
* ocamlPackages.ocamlgraph: 1.8.8 -> 2.0.0
* ocamlPackages.ocamlgraph_gtk: init 2.0.0

`ocamlgraph` upgrade is required for `dose3` package.

### Split off executables [editor](https://github.com/backtracking/ocamlgraph/blob/master/editor/dune) and [viewer](https://github.com/backtracking/ocamlgraph/blob/master/dgraph/dune)

Motivation: why would an application depending on this library have to inherit unused executables?

* Building executables ([editor](https://github.com/backtracking/ocamlgraph/blob/master/editor/dune), [viewer](https://github.com/backtracking/ocamlgraph/blob/master/dgraph/dune)) using buildDunePackage errors because it cannot find package, likely because it doesn't have `public_name` entry on dune file and also no opam package for application.

    Dune error is:
    > Error: I don't know about package dGraphViewer (passed through -p)

I don't use these executables and only needed to upgrade library for upgrading `dose3`.
* Possible solutions:
    1. ask upstream to add opam package and dune public_name entry to executables [[which I did](https://github.com/backtracking/ocamlgraph/issues/115)].
    2. patch it.
    3. do nothing.
      - It can be merged now.

### CODE REVIEW

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 126703` at f937e0e2 run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages failed to build:</summary>
  <ul>
    <li><a href="https://gist.github.com/6a554306d118a1a47baa2038c3e9ea86">dune-release</a></li>
    <li><a href="https://gist.github.com/b655c90bc460f2e9c3cd58291a169d4a">framac</a></li>
    <li><a href="https://gist.github.com/f513b10590ed9d3deda6ff4ceddc1e50">opa</a></li>
    <li><a href="https://gist.github.com/8cee99174e61dd9ed42b3a044e611f63">opam-installer</a></li>
  </ul>
</details>
<details>
  <summary>5 packages built successfully:</summary>
  <ul>
    <li>coqPackages.dpdgraph</li>
    <li>libbap</li>
    <li>python38Packages.bap</li>
    <li>python39Packages.bap</li>
    <li>why3</li>
  </ul>
</details>

Note that build failures [may predate](https://github.com/nix-community/hydra-check) this PR, and could be nondeterministic or hardware dependent.
Please exercise your independent judgement. Does something look off? Please file an [issue](https://github.com/rmcgibbo/r-rmcgibbo/issues) or reach out on IRC.

File: pkgs/development/ocaml-modules/ocamlgraph/gtk.nix [hunk 1]
```diff
@@ -0,0 +1,14 @@
+{ lib, callPackage, fetchurl, buildDunePackage, ocamlgraph, lablgtk, stdlib-shims, ... }:
```
Comment (vbgl) [inline pkgs/development/ocaml-modules/ocamlgraph/gtk.nix:1]:
```suggestion
{ buildDunePackage, ocamlgraph, lablgtk }:
```

File: pkgs/development/ocaml-modules/ocamlgraph/gtk.nix [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+buildDunePackage rec {
+  pname = "ocamlgraph_gtk";
+  inherit (callPackage ./. { }) version src useDune2 meta;
```
Comment (vbgl) [inline pkgs/development/ocaml-modules/ocamlgraph/gtk.nix:5]:
```suggestion
  inherit (ocamlgraph) version src useDune2 meta;
```

File: pkgs/development/ocaml-modules/ocamlgraph/gtk.nix [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+  buildInputs = [
+    ocamlgraph
+    lablgtk
+  ];
+  propagatedBuildInputs = [
+    stdlib-shims
+  ];
```
Comment (vbgl) [inline pkgs/development/ocaml-modules/ocamlgraph/gtk.nix:13]:
```suggestion
  propagatedBuildInputs = [
    ocamlgraph
    lablgtk
  ];
```

File: pkgs/development/ocaml-modules/ocamlgraph/default.nix [hunk 4]
```diff
@@ -1,39 +1,25 @@
...
-, gtkSupport ? true
-, lablgtk
-}:
+{ lib, fetchurl, buildDunePackage, stdlib-shims }:
 
-stdenv.mkDerivation rec {
+buildDunePackage rec {
   pname = "ocamlgraph";
-  version = "1.8.8";
+  version = "2.0.0";
```
Comment (vbgl) [inline pkgs/development/ocaml-modules/ocamlgraph/default.nix:5]:
```suggestion
  version = "2.0.0";
  
  minimalOCamlVersion = "4.03";
```

File: pkgs/development/ocaml-modules/ocamlgraph/gtk.nix (hunk 1 above)
Comment (superherointj) [reply pkgs/development/ocaml-modules/ocamlgraph/gtk.nix]: Done

File: pkgs/development/ocaml-modules/ocamlgraph/gtk.nix (hunk 2 above)
Comment (superherointj) [reply pkgs/development/ocaml-modules/ocamlgraph/gtk.nix]: Done

File: pkgs/development/ocaml-modules/ocamlgraph/gtk.nix (hunk 3 above)
Comment (superherointj) [reply pkgs/development/ocaml-modules/ocamlgraph/gtk.nix]: Done

File: pkgs/development/ocaml-modules/ocamlgraph/default.nix (hunk 4 above)
Comment (superherointj) [reply pkgs/development/ocaml-modules/ocamlgraph/default.nix:5]: Done

**Comment by vbgl** [comment]
This breaks Frama-C. Please add `stdlib-shims` to its run-time dependencies.

**Comment by superherointj** [comment]
> This breaks Frama-C. Please add `stdlib-shims` to its run-time dependencies.

Something odd is happening, `ocamlgraph` package is not exposing `stdlib-shims`. It is listed on `propagatedBuildInputs`.

$ nix build .#ocamlPackages.ocamlgraph
$ ls ./result/lib/ocaml/4.12.0/site-lib/
Only lists `ocamlgraph`.

Is this expected?

**Comment by vbgl** [comment]
> Is this expected?

Yes. What else should be there?

**Comment by superherointj** [comment]
> > Is this expected?
> 
> Yes. What else should be there?

I was expecting `stdlib-shims` to be there too.
I don't know why `framac` is not working.
Because `stdlib-shims` is already on buildInputs and being propagated.

**Comment by vbgl** [comment]
The wrapper for Frama-C sets the `OCAMLPATH` environment variable. Its content is explicitly written in the derivation, without relying on `buildInputs` or propagation or whatever. You have to add `stdlib-shims` there: https://github.com/NixOS/nixpkgs/blob/ad219bb3619ad7a7935b2cc1b1bf9bd4f2d28689/pkgs/development/tools/analysis/frama-c/default.nix#L17

I agree that it does not seem right. But this PR is not the right place to discuss how to fix the packaging of Frama-C.

File: pkgs/development/tools/analysis/frama-c/default.nix [hunk 5]
```diff
@@ -38,7 +39,7 @@ stdenv.mkDerivation rec {
 
   buildInputs = with ocamlPackages; [
     ncurses ocaml findlib ltl2ba ocamlgraph yojson menhirLib camlzip
-    lablgtk coq graphviz zarith apron why3 mlgmpidl doxygen
+    lablgtk coq graphviz stdlib-shims zarith apron why3 mlgmpidl doxygen
```
Comment (vbgl) [inline pkgs/development/tools/analysis/frama-c/default.nix:42]: This one is not needed.

Comment (superherointj) [reply pkgs/development/tools/analysis/frama-c/default.nix]: Done

**Comment by vbgl** [comment]
Frama-C also depends (at build-time only) on `ocamlgraph_gtk` for showing call-graphs.

**Pull request merged**